enhancements/validate-catalog-for-bad-bundles.md

```diff
@@ -0,0 +1,177 @@
+
+## Part 1: Validate bundle objects
+
+### 1. Extending Validation library to add validations in `Golang`
+- Extend [BundleValidator](https://github.com/operator-framework/api/blob/master/pkg/validation/internal/bundle.go#L14) to include new validation functions for each of the SAP Gardener objects. (ToDo: Do we add any checks under [BundleFormat](https://github.com/operator-framework/operator-registry/blob/master/pkg/lib/bundle/validate.go#L65) ?)
+- The validation function for each kube manifest object would check for conditions specified under [Static Validation Rules](https://hackmd.io/7TlDq9t_RuGjGm8YsEbyHA?both#Static-Validation-Rules).
+- (ToDo: 
+    -    Do we want to maintain some sort of `struct` for storing various kube-manifest objects as these objects could grow (in the future) or
+    -    Wait to see how these new objects are represented in the [Bundle](https://github.com/operator-framework/api/blob/master/pkg/manifests/bundle.go#L11) as part SAP gardener PR and then take a call)
+ 
+### 2. Using [CUE](https://cuelang.org/docs/integrations/yaml/#validate-yaml-files)
+- CUE is a configuration language that allows you to define a `<pdb-checks>.cue` template for validating Kube Manifests like PDBs.
+- CUE can be an extendable validation library to accept validate template for a resource like PDBs before applying them on to a cluster.
+
+
+#### Pros
+- Not hinging on a third-party operator like OPA Gatekeeper.
+- Validation is extendable to customized .cue files
+
+#### Cons
+- Not a lot to offer in terms of flexibility as compared to [OpenPolicyAgent (OPA)](https://kubernetes.io/blog/2019/08/06/opa-gatekeeper-policy-and-governance-for-kubernetes/)
+- Require users supplying CUE validation template to be familiar with a new type of language
+
+### 3. Using [KUBEVAL](https://kubeval.instrumenta.dev/) - (yet to investigate)
+
+
+## Part 2: Tool to list "bad" bundles in a catalog
+The basic approach will be to:
+- Extend [`opm`](https://github.com/operator-framework/operator-registry/blob/master/docs/design/opm-tooling.md) tooling to add a new command `opm registry extract`
+- Pull the catalog image referenced by `--index` to extract the database file and store it in a location specified by `--download-folder`
+- Read the database file to get the list of bundle images
+- Reuse `opm alpha bundle validate` functionality to validate each bundle and list out the "bad" ones
+
+### Extract the contents of a catalog source image 
+```shell
+opm registry extract --index quay.io/operator-framework/upstream-community-operators:latest --download-folder ./index-manifest
+```
+
+- `--index` flag references the catalog source image which has the running index image
+- `--download-folder` flag points to a directory location where the extracted file(s) gets stored
+
+### Define a new API 
+- Define a struct that stores the command line arguments from the `opm registry extract` command. 
```

Comment:
This seems like a really useful API in general, even outside the scope of validation. Maybe we should separate these out so that there is a way to pull down and unpack the database as a command, and then have a second command that did the validation? Or perhaps a command that wrapped up both of those APIs if you wanted to do it in one go?

Reply:
I have split the command into two parts:
`opm index extract <catalog-image>` - pull and unpack catalog image
`opm index validate <registry file location>` - validate all the bundle images in the registry file